sys/include/net/nanocoap_sock.h

```diff
@@ -277,6 +273,24 @@ ssize_t nanocoap_get_blockwise_url_to_buf(const char *url,
  */
 ssize_t nanocoap_sock_request(nanocoap_sock_t *sock, coap_pkt_t *pkt, size_t len);
 
+/**
+ * @brief   Simple synchronous CoAP request with callback
+ *
+ *          The response will be handled by a callback, which avoids copying the
+ *          response packet out of the network stack internal buffer.
+ *
+ * @param[in]       sock    socket to use for the request
+ * @param[in,out]   pkt     Packet struct containing the request. Is reused for
+ *                          the response
+ * @param[in]       cb      Callback executed for response packet
+ * @param[in]       arg     Optional callback argumnent
+ *
+ * @returns     length of response on success
+ * @returns     <0 on error
+ */
+ssize_t nanocoap_sock_request_cb(sock_udp_t *sock, coap_pkt_t *pkt,
+                                 coap_request_cb_t cb, void *arg);
```

Comment:
In the implementation I can see that the response callback is not executed when the response has no payload. I'm not sure if that is a good idea, a caller likely still wants to evaluate the status code of the response.

E.g. consider a `2.03 Valid` reply without payload that contains an `Etag` Option. In this case the response would have a payload from a high level point of view: The client included one or more Etag Options in the request to tell the server which states of the resource the client has a cached payload for, so that the server can skip transmitting the payload and just tell which of the provided Etags is now valid.